### PR TITLE
DM-12170: First draft

### PR DESCRIPTION
**Pull request by timj**

### CODE REVIEW

**Comment by r-owen** [comment]
Substantive points

5.3c "These value added repositories then need to be shareable to the commissioning archive and potentially to QA collaborators.". What does "shareable to the commissioning archive" mean? Made visible to the commissioning archive but kept where they are? Able to be ingested into the commissioning archive? I suggest spelling this out, if we know the answer, as it seems to me readers may come up with very different interpretations.

I reworded some alert production use cases and some requirements according to K-T's feedback but have not heard back from him. I just pinged him, asking him to take another look. This document does include the latest versions (aside from some minor fixes noted below), but if he wants changes, we will need to accommodate.

Picky points:

Needs resolution: Data Facility Scientist (leadership?)

Please define PVI, LSP, DAX, EFD. Simply putting "DAX" in parenthesis in the "Data Access" section title would probably suffice for that one.

OCS, EPO and DAC are used so rarely I suggest we spell them out, especially in text (titles are less important, especially if the acroynym is spelled out shortly after the title).

When an actor is named I suggest appear in capital letters. There is a mix.

I'm grumpy about the extensive use of "they" to refer to a single human. It's probably not worth fixing, but I was heartened by the use of a mix of "he" and "she" in the DRP use cases, and some use of plurals (making "they" appropriate) in some earlier use cases.

Typo: perofmrnace in "Validation Scientist Science Operations Scientist (long term perofmrnace)"

3.1 extra "to" in "may require the ability to for SuperTasks to cooperatively run in parallel".
I fixed the spreadsheet.

3.1e: The spreadsheet had some grammatical errors. I have fixed those. Please copy again.

3.2a: typo: extra period in "all will be processed that way. when". Fixed on spreadsheet.

3.2a "The job is managed differently; I believe it will be triggered manually by an Observer"
should be cleaned up:
- Remove the "I" and reword.
- We need to state who really will trigger it. An Observer sounds correct, but if it's more automated the Batch Production Operator?

3.6a "A pipelines developer is writing ...": capitalize Pipeline Developer?

4.1a Run-on sentence: "A Developer wants to read in some raw data files in FITS format and do test processing runs writing out FITS in one case and HDF5 in the other in order to compare performance differences.".

4.2b "An afterburner is used when processing is complete that reads in both forms of the output for a single dataset type, and they are analyzed to confirm that they contain the same information.". I suggest active voice (e.g. "Run an afterburner when..." and breaking it into two sentences. It's not clear to mea what "they" refers to in "they are analyzed".

4.3a "and a fully-populated data header for each visit using updated values": comma between "visit" and "using" to make it clear the text applies to each visit, not just those visit that use updated values.

4.4b "when they are signalled that the jobs have been completed" I suggest "informed" instead of "signalled".

4.5a "An Astronomer downloads a coadd from the Data Access Center portal that comes with full metadata (they do not know this might be a composite dataset in the DAC)." I find the parenthetical comment unhelpful and confusing. I strongly suggest omitting it. What is the point? That the should not *have* to know that it was saves as a composite?

4.5a "They want to see how it changed from the version they have from the previous data release which they were using for their research paper". I find this very confusing and suggest rewording. Perhaps "They want to see if the current version of the coadd has changed compared to the version they are using, which came from an older data release".

5.9a "The commissioning scientists will need ...to compare to.". Perhaps "...compare against" or "for comparisons"?

5.12a "will need to spawn on the fly processing". For clarity I suggest on-the-fly or putting "on the fly" in quotes.

5.12a "This processing may require both one off investigatory type and more formal batch style". Please reword.

5.13a "...on many different cadences: inter-night, daily, monthly, on release cadence." I suggest replacing "on release cadence" (which I find confusing) with something like "per data release". I am also not entirely sure what was intended by "inter-night". Taken at face value I believe it means "between nights", which is approximately the same as "daily" and could thus be omitted! If it means "during the night" then I suggest saying that, instead.

6.2a "to produce measurements of various metrics: both technical like runtime and more algorithm related like number of sources.". Could use some commas.

7.6b "In order to prevent hits to network disks" should probably read "In order to reduce reading network disks", though that begs the question of where the cache will be stored. If we know (memory? non-network disks?) then I suggest saying so.

7.6c I am not sure of the point of this one. Perhaps it is saying that DAX augments headers of images when it reads images, and that the cache should include those augmented headers?

7.8b Do we *know* that the images will end up in a single directory with no structure? Or are you saying that they *may* end up that way? I was personally hoping for some structure.


8.1a "The input files that are required for the job to compute must be determined": I suggest "...required for the job to execute...".

8.1b "The relevant files may be pre-staged to a filesystem closer to the compute site and the job is submitted to the batch system.". This seems to be two barely related thoughts. I suggest ditching "and the job..." or at least making it two sentences.

8.1c "When the job completes, the job’s file system the expected output files that have been configured to be saved are either stored in ...". Please reword!

8.5 is a duplicate of 3.1

8.6c "and runs the same notebook submitting batch jobs". "and run"? Add a comma: "the same notebook, submitting batch jobs".

8.6c "They do not edit his notebook to indicate..." Should it be "this notebook"?

8.7a "They run a program which deletes the file(s) from disk and updates the local repository such that the file is no longer discovered." I think "dataset is no longer discoverable" would be clearer. Or if "file" is preferred, then "the file(s) are no longer discoverable".

8.8c "This information is flagged as suspect" -> "This observation is flagged as suspect" or, better yet: "The scientist reports this observation as suspect, using a standard tool and providing supporting evidence. The Data Release Board examines the report, agrees with the assessment, and..."

8.8d "the same flagging process can be told to also flag/change data status all data products that have contributions from that observation" At least one word is missing. I suggest this simpler version: "can be told to flag all data products that have contributions from that observation"

8.9a "They then override the bad observation filtering and retrieve". I found this confusing (is the observation filtering defective?), though I was able to work it out. Consider "They then enable inclusion of bad observations, and retrieve".

8.11a "When all the jobs are completed the output products", comma after "completed".
8.11b "A QA job performs": does the Production Operator manually trigger this after all the jobs are completed? Does it somehow get triggered automatically (as the current text vaguely implies). If we know, I suggest spelling it out.

9.2b "but the dataset is predefined" -> "...is a predefined"

9.2c an example would help, such as retrieving an exposure with a WCS that was refined by additional processing after the exposure was written.

9.4 I find this very unclear. I am *guessing* that the point to run a processing SuperTask and not save some specified datasets (typically intermediate products) in order to save storage space. It would help to start with a summary of what's going on before saying "Like DRP...".

9.6a picky, but I'd ditch the parenthesis in "transfers output...to (more) permanent storage".

9.7a "and identifying raw data and sky patches from each other based on location". What does the "from each other" mean? Are you saying we should be able to look this up on either direction?

9.2 and 9.8 are basically identical and have far too much duplication. I suggest combining them, if at all possible. If that is not possible, then consider rewriting them as follows:
- Write one as a use case plus a list of variants. Preferably use no letter for the main use case text and use a lettered list or bullet list for for the variants.
- Write the other as a single use case and point to the first use case for variants.

9.13 Interesting. I hadn't realized this would be necessary.

**Comment by TallJimbo** [comment]
> 9.4 I find this very unclear. I am guessing that the point to run a processing SuperTask and not save some specified datasets (typically intermediate products) in order to save storage space. It would help to start with a summary of what's going on before saying "Like DRP...".

Yes, this is what is described here (though saving on I/O time is probably more important than saving on storage space, since we can also choose to remove/not transfer data products later).

However, I disagree philosophically with the idea that we should repeat the text of DRP2 and DRP3 here; I think precision and maintainability (i.e. "text reuse") is more important than readability here.  But I'll leave the decision on that up to @timj, as he knows what use cases are supposed to look like more than I do.

> 9.7a "and identifying raw data and sky patches from each other based on location". What does the "from each other" mean? Are you saying we should be able to look this up on either direction?

Yes, we should be able to look up raw data overlapping a set of sky patches and look up the sky patches overlapping some raw data.  Could perhaps rewrite as "looking up raw data that overlap sky patches or vice versa".

> 9.2 and 9.8 are basically identical and have far too much duplication. I suggest combining them, if at all possible.

I think their similarity is a hint towards the desired interfaces, but I think that they are conceptually different use cases, given that they have different actors, different environments, and potentially different information provided by the user to the data access system.  If @timj has an easy way to make them share text, I'm all for it, but it's more important that we permit requirements or design components to tick these off separately.

File: LDM-592.tex [hunk 1]
```diff
@@ -0,0 +1,1200 @@
...
+
+\item
+They have another collection that is from the previous version but this one can be read by the current library, although a warning is issued that they may want to migrate to make use of new features.
+
+}
+
+\usecase{DAX9}{Using the Butler with VO retrieved data originating from butler repositories}{%
+
+\item
+An Astronomer downloads images from the LSST DAC using the VO.
```
Comment (brianv0) [inline LDM-592.tex]: -> "using the VO services."

File: LDM-592.tex [hunk 2]
```diff
@@ -0,0 +1,1200 @@
...
+A Staff Scientist has a data collection that is too old to be read by the current version of the software.
+
+\item
+The software they are using reports the version number and suggests they run the migration script.
+
+\item
+Once the migration script completes the data collection can be read.
+
+\item
+They have another collection that is from the previous version but this one can be read by the current library, although a warning is issued that they may want to migrate to make use of new features.
```
Comment (brianv0) [inline LDM-592.tex]: Maybe the following: "They have another collection that is from the previous version but this one which is still supported by the current library,"

**Comment by r-owen** [comment]
More comments (this gets me through the 10.x use cases)

9.14 "via a set of processing runs runs (possibly in a different environment)":
- what does "possibly in a different environment" mean? No environment has been stated yet. I suggest deleting or rewriting that bit.
- remove one of the repeated "runs"

9.15 "pre-cache some data, from a remote repository" I suggest removing the comma

9.16 "It would be efficient if this can be cached" fix the grammar?
Consider rewriting to make it more like a use case. "It would be efficient" sounds more like a "nice to have" requirement.

9.17 Please flesh this out. Does the astronomer want to operate on the whole dataset at once (e.g. on one node, in one task), or on a piece small enough to fit into memory? Those are entirely different uses cases, and I hope we don't have to support the first one!

9.18 is a requirement, not a use case. Is it worth trying to reword it? A bit of justification seems in order.

The 10.x use cases appear to have originally been written as an overall paragraph followed by a list of steps. Unfortunately the introduction firs step seem to have gotten combined. This would be easy to fix.

10.2 picks up where 10.1 leaves off. Please make it stand on its own, e.g.: "A Science Validation Collaborator finds a particular calibrated exposure that messed up a coadd (see 10.1). ..."

10.4 starts out too wordy. I'm also not happy with bringing in a "job object", since that is an implementation detail.  Consider "A developer instruments an image characterization Task in order to measure a specific set of metrics about the Task. The developer wants to publish this set of metrics to SQuaSH, using a descriptive name."

10.7 "so that they can do comparisons of how the change in configuration affects the output". I suggest "so they can determine how the change ..."

10.12 I'm not convinced this is needed. Patches are intended to be small enough to not need further subdivision.

File: LDM-592.tex [hunk 3]
```diff
@@ -0,0 +1,1202 @@
...
+When values from nightly processing are not available (e.g. in construction, or when testing a new deployment of the operations system), a Staff Scientist sets (or resets) the metadata values from an initial manually-invoked processing run that does not extend to coaddition.
+}
+
+\usecase{DRP15}{Master Calibration Products Registration}{%
+
+\item
+A Staff Scientist produces a set of master calibration products via a set of processing runs runs (possibly in a different environment) and adds them to a data repository for use in other processing runs.
+Old master calibration products should remain available, but one or more "preferred" labels may need to be transferred to the new products.
+When the data repository is part of the operations system, this is done by a Batch Production Operator, not arbitrary Staff Scientists.
+}
```
Comment (TallJimbo) [inline LDM-592.tex]: New text:
```
A Staff Scientist produces a set of master calibration products via a set of processing runs runs (possibly in a different environment) and adds them to a data repository for use in other processing runs.
Old master calibration products should remain available, but one or more "preferred" labels may need to be transferred to the new products.
Each set of master calibration products includes data products with (date or visit) validity ranges; selecting the set of calibrations to be used for a processing run is distinct from selecting the data product within that set to use for a particular observation.
When the data repository is part of the operations system, this is done by a Batch Production Operator, not arbitrary Staff Scientists.
```

File: LDM-592.tex [hunk 4]
```diff
@@ -0,0 +1,1202 @@
...
+A Staff Scientist or Astronomer has existing code (written to use \texttt{butler.get()} and \texttt{butler.put()} calls) that they wish to use with no (or minimal) changes.
+}
+
+\usecase{DRP29}{Multi-Repository File Removal}{%
+
+\item
+A shared storage space used for collaborative algorithm development and testing is filling up, and developers respond by removing processing outputs that are not being used anymore.
+Some of the processing runs to be deleted were based on inputs produced in other processing runs that should not be deleted, and those inputs are hence considered to be part of both the to-be-deleted data repositories and the to-be-kept data repositories.
+After removing the to-be-deleted data repositories, the files that are no longer used by any repository are identified and deleted, while those in the to-be-kept repositories are not.
+}
```
Comment (TallJimbo) [inline LDM-592.tex:947]: Text for a new use-case:
```
\usecase{DRP30}{Forced deletion of datasets}

\item
One or more data products are deleted by a production operator or other authorized actor, without regard for which data repositories or other users may refer to them.  In some contexts this may not remove all record of these datasets held outside their primary storage locations, as long as the storage cost of those records is negligible.
```

File: LDM-592.tex [hunk 5]
```diff
@@ -171,15 +171,15 @@ \section{Alert Production}
 The user, a Staff Scientist, specifies a visit to be processed.
 
 \item
-The butler expands the incomplete data ID by setting the filter and selecting only those CCDs for which raw image data is available.
+The butler expands the incomplete DatasetExpresion by setting the filter and selecting only those CCDs for which raw image data is available.
```
Comment (pschella) [inline LDM-592.tex]: Not sure a `DatasetExpression` can be incomplete at all (apart from a list of `DatasetType`s), since it represents a SQL query.

File: LDM-592.tex [hunk 6]
```diff
@@ -1212,6 +1212,61 @@ \section{Science Validation}
 
 \bibliography{local,lsst,lsst-dm,refs,books,refs_ads}
 
+\section{Glossary}
+
+\begin{description}[style=nextline]
+
+\item[Dataset] Represents a single discrete entity of data, with associated metadata (e.g.\ a particular \texttt{calexp} for a particular instrument corresponding to a particular visit and sensor).
+
+\item[ConcreteDataset] The in-memory manifestation of a Dataset (e.g. an \texttt{afw::image::Exposure} instance with the contents of a particular \texttt{calexp}).
```
Comment (pschella) [inline LDM-592.tex]: Now called `InMemoryDataset`.

File: LDM-592.tex [hunk 7]
```diff
@@ -1212,6 +1212,61 @@ \section{Science Validation}
...
+
+\item[DatasetType] The conceptual type of which Datasets are instances (e.g.\ \texttt{calexp}).  A new DatasetType can be defined by combining a DatasetMetatype, a DatasetIdentifierType, and a string name.
+
+\item[DatasetMetatype] A category of DatasetTypes that utilize the same in-memory classes for their ConcreteDatasets and can be saved to the same file format(s), using the same low-level input and output code.
```
Comment (pschella) [inline LDM-592.tex]: Now called `StorageClass`.

File: LDM-592.tex [hunk 8]
```diff
@@ -171,15 +171,15 @@ \section{Alert Production}
 The user, a Staff Scientist, specifies a visit to be processed.
 
 \item
-The butler expands the incomplete DatasetExpresion by setting the filter and selecting only those CCDs for which raw image data is available.
+The butler expands the DatasetExpresion by setting the filter and selecting only those CCDs for which raw image data is available.
```
Comment (pschella) [inline LDM-592.tex]: I'd prefer: "The butler expands the DatasetExpresion to select only those CCDs for which raw image data is available."
